python/cudf/cudf/core/_internals/where.py

```diff
@@ -0,0 +1,395 @@
+        return source_col, other_scalar
+
+    device_scalar = _normalize_scalars(source_col, other_scalar)
+
+    if other_scalar is None:
+        return source_col, device_scalar
+    elif cudf.utils.dtypes.is_mixed_with_object_dtype(
+        device_scalar, source_col
+    ):
+        raise TypeError(
+            "cudf does not support mixed types, please type-cast "
+            "the column of dataframe/series and other "
+            "to same dtypes."
+        )
+    if inplace:
+        if not cudf.utils.dtypes.can_cast(
+            device_scalar.dtype, source_col.dtype
+        ):
+            warnings.warn(
+                f"Type-casting from {device_scalar.dtype} "
+                f"to {source_col.dtype}, there could be potential data loss"
+            )
+        return source_col, device_scalar.astype(source_col.dtype)
+    else:
+        if pd.api.types.is_numeric_dtype(
+            source_col.dtype
+        ) and cudf.utils.dtypes.can_cast(other_scalar, source_col.dtype):
+            common_dtype = source_col.dtype
+        else:
+            common_dtype = cudf.utils.dtypes.find_common_type(
+                [source_col.dtype, np.min_scalar_type(other_scalar)]
+            )
+
+        source_col = source_col.astype(common_dtype)
+        return source_col, cudf.Scalar(other_scalar, dtype=common_dtype)
+
+
+def _normalize_columns_and_scalars_type(
+    frame: Union[Series, Index, DataFrame], other: Any, inplace: bool = False,
+) -> Tuple[
+    Union[Series, Index, DataFrame, ColumnLike], Any,
+]:
+    """
+    Try to normalize the other's dtypes as per frame.
+
+    Parameters
+    ----------
+
+    frame : Can be a DataFrame or Series or Index
+    other : Can be a DataFrame, Series, Index, Array
+        like object or a scalar value
+
+        if frame is DataFrame, other can be only a
+        scalar or array like with size of number of columns
+        in DataFrame or a DataFrame with same dimension
+
+        if frame is Series, other can be only a scalar or
+        a series like with same length as frame
+
+    Returns:
+    --------
+    A dataframe/series/list/scalar form of normalized other
+    """
+    if isinstance(frame, DataFrame) and isinstance(other, DataFrame):
+        source_df = frame.copy(deep=False)
+        other_df = other.copy(deep=False)
+        for self_col in source_df._data.names:
+            source_col, other_col = _check_and_cast_columns(
+                source_col=source_df._data[self_col],
+                other_col=other_df._data[self_col],
+                inplace=inplace,
+            )
+            source_df._data[self_col] = source_col
+            other_df._data[self_col] = other_col
+        return source_df, other_df
+
+    elif isinstance(
+        frame, (Series, Index)
+    ) and not cudf.utils.dtypes.is_scalar(other):
+        other = cudf.core.column.as_column(other)
+        input_col = frame._data[frame.name]
```

Comment:
This won't work if `frame` is a `MultiIndex` which is an instance of `Index`

Reply:
So we(& pandas) error when `MultiIndex.where` is invoked, so this code-flow won't be hit.